File: codegen/features/xsls/backendless-codegen-xsls/ios/backendless-codegen/ObjC/Chat/Chat/ChatViewController.m
#import "ChatViewController.h"

@interface ChatViewController() {
    UITextField *activeTextField;
}
@end

@implementation ChatViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.title = self.userName;
    self.chatTextView.delegate = self;
    self.messageTextField.delegate = self;
    [self addMessageListener];
    [self publishMessage:[NSString stringWithFormat:@"%@ joined", self.userName]];
    [self.navigationItem setHidesBackButton:YES];
    [self.navigationItem setBackBarButtonItem:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardDidShow:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillBeHidden:) name:UIKeyboardWillHideNotification object:nil];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIKeyboardWillHideNotification object:nil];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    [self.chatTextView scrollRangeToVisible:NSMakeRange([self.chatTextView.text length] - 1, 1)];
}

- (void)addMessageListener {
    __weak ChatViewController *weakSelf = self;
    RTSubscription *subscription __unused = [self.channel addStringMessageListenerWithResponseHandler:^(NSString *message) {
        if ([self.chatTextView.text length] > 0) {
            [weakSelf.chatTextView insertText:[NSString stringWithFormat:@"\n\n%@", message]];
        }
        else {
            [weakSelf.chatTextView insertText:[NSString stringWithFormat:@"%@", message]];
        }
    } errorHandler:^(Fault *fault) {
        [weakSelf showErrorAlert:fault];
    }];
}

- (void)publishMessage:(NSString *)message {
    [Backendless.shared.messaging publishWithChannelName:self.channel.channelName message:message responseHandler:^(MessageStatus *messageStatus) {
    } errorHandler:^(Fault *fault) {
        [self showErrorAlert:fault];
    }];
}

- (void)keyboardDidShow:(NSNotification *)notification {
    NSValue *infoValue = [notification.userInfo valueForKey:UIKeyboardFrameEndUserInfoKey];
    CGSize keyboardSize = infoValue.CGRectValue.size;
    [UIView animateWithDuration:0.3 animations:^{
        CGRect viewFrame = self.view.frame;
        viewFrame.size.height -= keyboardSize.height;
        self.view.frame = viewFrame;
    }];
}

-(void)keyboardWillBeHidden:(NSNotification *)notification {
    [UIView animateWithDuration:0.3 animations:^{
        CGRect screenFrame = [[UIScreen mainScreen] bounds];
        CGRect viewFrame = CGRectMake(0, 0, screenFrame.size.width, screenFrame.size.height);
        viewFrame.origin.y = 0;
        self.view.frame = viewFrame;
    }];
}

- (void) textFieldDidBeginEditing:(UITextField *)textField {
    activeTextField = textField;
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    [textField resignFirstResponder];
    activeTextField = nil;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if ([self.messageTextField.text length] > 0) {
        [self publishMessage:[NSString stringWithFormat:@"[%@]: %@", self.userName, self.messageTextField.text]];
        self.messageTextField.text = @"";
    }
    [textField resignFirstResponder];
    return YES;
}

- (void)showErrorAlert:(Fault *)fault {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:[NSString stringWithFormat:@"Error %li", fault.faultCode] message:fault.message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *dismissButton = [UIAlertAction actionWithTitle:@"Dismiss" style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:dismissButton];
    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: codegen/features/xsls/backendless-codegen-xsls/ios/backendless-codegen/Swift/Chat/Chat/ChatViewController.swift

import UIKit
import Backendless

class ChatViewController: UIViewController, UITextFieldDelegate, UITextViewDelegate {
    
    @IBOutlet var chatTextView: UITextView!
    @IBOutlet var messageTextField: UITextField!
    
    var userName: String!
    var channel: Channel!
    var activeTextField: UITextField?

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = userName
        chatTextView.delegate = self
        messageTextField.delegate = self
        addMessageListener()
        publishMessage(String(format: "%@ joined", userName!))
        navigationItem.hidesBackButton = true
        navigationItem.backBarButtonItem = nil
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardDidShow(_:)), name: UIResponder.keyboardWillShowNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillBeHidden(_:)), name: UIResponder.keyboardWillHideNotification, object: nil)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        NotificationCenter.default.removeObserver(self, name: UIResponder.keyboardWillShowNotification, object: nil)
        NotificationCenter.default.removeObserver(self, name: UIResponder.keyboardWillHideNotification, object: nil)
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        chatTextView.scrollRangeToVisible(NSRange(location: chatTextView.text.count - 1, length: 1))
    }
    
    func addMessageListener() {
        let _ = channel.addStringMessageListener(responseHandler: { message in
            if (!self.chatTextView.text.isEmpty) {
                self.chatTextView.insertText(String(format: "\n\n%@", message))
            }
            else {
                self.chatTextView.insertText(String(format: "%@", message))
            }
        }, errorHandler: { fault in
            self.showErrorAlert(fault)
        })
    }
    
    func publishMessage(_ message: String) {
        Backendless.shared.messaging.publish(channelName: channel.channelName, message: message, responseHandler: { messageStatus in
        }, errorHandler: { fault in
            self.showErrorAlert(fault)
        })
    }
    
    @objc func keyboardDidShow(_ notification: NSNotification) {
        let infoValue = notification.userInfo![UIResponder.keyboardFrameEndUserInfoKey] as! NSValue
        let keyboardSize = infoValue.cgRectValue.size
        UIView.animate(withDuration: 0.3, animations: {
            var viewFrame = self.view.frame
            viewFrame.size.height -= keyboardSize.height
            self.view.frame = viewFrame
        })
    }
    
    @objc func keyboardWillBeHidden(_ notification: NSNotification) {
        UIView.animate(withDuration: 0.3, animations: {
            let screenFrame = UIScreen.main.bounds
            var viewFrame = CGRect(x: 0, y: 0, width: screenFrame.size.width, height: screenFrame.size.height)
            viewFrame.origin.y = 0
            self.view.frame = viewFrame
        })
    }
    
    func textFieldDidBeginEditing(_ textField: UITextField) {
        activeTextField = textField
    }
    
    func textFieldDidEndEditing(_ textField: UITextField) {
        textField.resignFirstResponder()
        activeTextField = nil
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        if (!(messageTextField.text?.isEmpty)!) {
            publishMessage(String(format: "[%@]: %@", userName!, messageTextField.text!))
            messageTextField.text = ""
        }
        textField.resignFirstResponder()
        return true
    }
    
    func showErrorAlert(_ fault: Fault) {
        let alert = UIAlertController(title: String(format: "Error %@", fault.faultCode), message: fault.message, preferredStyle: .alert)
        let dismissButton = UIAlertAction(title: "Dismiss", style: .cancel, handler: nil)
        alert.addAction(dismissButton)
        present(alert, animated: true, completion: nil)
    }
}
